### PR TITLE
Auto detect DynamicDataSourceType

### LINKED ISSUE #4254: Simplify `[DynamicData]` to not require user to specify the member type
## Background and Motivation

Reflection call to detect if the member is property or method isn't expensive so there is no good reason to force user to have to specify member type. Besides, nowadays most people will use `nameof(MemberName)` that help to guarantee the member exists.

## Proposed Feature

I think the easiest and most transparent solution is probably to add a new state to the enum `DynamicDataSourceType` like `AutoDetect` that would be the default value. Although this is technically a breaking change, it should not have any impact for users who are using the default value or the explicit value.

Note we need to update the related code analyzer.

## Alternative Designs

Deprecate enum and related ctors but that would flow to the user for no real impact.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #4254

### CODE REVIEW

File: src/Adapter/MSTest.TestAdapter/DynamicDataOperations.cs [hunk 1]
```diff
@@ -25,36 +25,33 @@ public IEnumerable<object[]> GetData(Type? _dynamicDataDeclaringType, DynamicDat
...
+                {
+                    obj = GetDataFromProperty(dynamicDataPropertyInfo);
+                }
+                else if (PlatformServiceProvider.Instance.ReflectionOperations.GetDeclaredMethod(_dynamicDataDeclaringType, _dynamicDataSourceName) is { } dynamicDataMethodInfo)
+                {
+                    obj = GetDataFromMethod(dynamicDataMethodInfo);
+                }
+                else
+                {
+                    // throw
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/DynamicDataOperations.cs]: Yes let's throw `NotSupportedException`. I know some other frameworks do support fields but I have never really found any good use case for it so not sure if that's worth supporting.

@nohwnd any opinion?

**Review by Evangelink — DISMISSED** [review]
Some issue with public API it seems but otherwise LGTM

File: src/Adapter/MSTest.TestAdapter/DynamicDataOperations.cs (hunk 1 above)
Comment (nohwnd) [reply src/Adapter/MSTest.TestAdapter/DynamicDataOperations.cs]: properties and methods seem fine if that is the current thing we support in mstest.

Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/DynamicDataOperations.cs]: @Youssef1313 please add the throw here

Comment (Youssef1313) [reply src/Adapter/MSTest.TestAdapter/DynamicDataOperations.cs]: In progress

**Review by Evangelink — APPROVED** [review]

**Pull request merged**